mesh/mesh.go

```diff
@@ -500,3 +500,66 @@ func (m *Mesh) GetATXs(atxIds []types.AtxId) (map[types.AtxId]*types.ActivationT
 	}
 	return atxs, mIds
 }
+
+// ActiveSetForLayerView - returns the active set size that matches the view of the contextually valid blocks in the provided layer
+func (m *Mesh) ActiveSetForLayerView(layer types.LayerID, layersPerEpoch uint16) (uint32, error) {
+
+	epoch := layer.GetEpoch(layersPerEpoch)
+	firstLayerOfPrevEpoch := types.LayerID(epoch-1) * types.LayerID(layersPerEpoch)
+
+	// build a map of all blocks between firstLayerOfPrevEpoch and our current layer
+	mp := make(map[types.BlockID]struct{})
+	blocks, err := m.LayerBlockIds(layer)
+	for _, bid := range blocks {
+		mp[bid] = struct{}{}
+	}
+
+	activeSetSize := uint32(0)
+	countedAtxs := make(map[string]types.AtxId)
+
+	traversalFunc := func(blkh *types.BlockHeader) error {
+
+		// only contextually valid blocks
+		if !m.IsContexuallyValid(blkh.Id) {
+			return nil
+		}
+
+		blk, err := m.GetBlock(blkh.Id)
+		if err != nil {
+			return err
+		}
+
+		// count unique ATXs
+		for _, id := range blk.AtxIds {
+			atx, err := m.GetAtx(id)
+			if err != nil {
+				log.Panic("error fetching atx %v from database -- inconsistent state", id.ShortId()) // TODO: handle inconsistent state
```

Comment:
remove panic

Reply:
Are you sure about it?
I copied it from atxdb. If we can't get the ATX at this point it means we didn't keep the atx which we said we kept.